### ISSUE TITLE
Module not loaded properly when file and folder have same name

### ISSUE THREAD

**Issue by cristi-badila** [systemjs/systemjs#1094]
First of all I would like to state that I can't say for sure if this is a known/design limitation or a real bug but to me it seems like a bug.

Given the following folder structure:
```
project_name
|
---index.html
---node_modules
---app
------components.js
------components
---------layout
------------shell.js
```

and the following configuration
```
    System.config({
      packages: {
        'app': {
          format: 'register',
          map: {
            "./components": "/app/components.js",
            "./components/layout/shell": "/app/components/layout/shell.js"
          }
        }
      }
    });
```

and the following ``components`` module
```
System.register(['./components/layout/shell'], function(exports_1) {
    "use strict";
    return {
        setters:[
          function (app_1_1) {
                exports_1({
                    "Shell": app_1_1["Shell"]
                });
            }],
        execute: function() {
        }
    }
});
```

executing the following code:
```
    System.import('/app/components');
```
results in SystemJS trying to load the ``Shell`` module from a path similar to ``http://localhost/app/components.js/layout/shell.js`` instead of ``http://localhost/app/layout/shell.js``

You can find a sample app [here](https://github.com/cristi-badila/systemjsissue) to reproduce the issue.

To me it seems that this is a bug that stems from the implementation of ``getMapMatch`` but I can't say for sure as I don't know this project too well. 

The version of SystemJS that I've used is ``0.19.20``

**Comment by guybedford** [comment]
This is actually correct. When importing `app/components/layout/shell` that matches the map configuration that says it should treat `./components/layout/shell` as `app/components/layout/shell.js`. This package contextual map then runs through the top-level normalization again. So we treat that as if importing `app/components/layout/shell.js` this then matches the `app/components` map as map matches subpaths as well. As a result it becomes `app/components.js/layout/shell.js`.

Firstly, note that relative maps can be used to map to paths within the package:

```javascript
    System.config({
      packages: {
        'app': {
          format: 'register',
          map: {
            "./components": "./components.js",
            "./components/layout/shell": "./components/layout/shell.js"
          }
        }
      }
    });
```

When using relative paths like this is won't go through the top-level normalize again so that then avoids the double mapping problem.

**Comment by cristi-badila** [comment]
@guybedford thanks for the clarification. I didn't realize the fact about relative paths as there were a lot of calls being done in order to resolve/normalize the module names and it was kind of hard to follow everything. Thank you again.
